velox/functions/prestosql/tests/DateTimeFunctionsTest.cpp

```diff
@@ -674,6 +674,236 @@ TEST_F(DateTimeFunctionsTest, dateTruncDate) {
+  EXPECT_EQ(std::nullopt, dateAdd("month", std::nullopt, Timestamp(0, 0)));
+
+  // Check invalid units
+  EXPECT_THROW(dateAdd("invalid_unit", 1, Timestamp(0, 0)), VeloxUserError);
+
+  // Simple tests
+  EXPECT_EQ(
+      Timestamp(1551348061, 999'999) /*2019-02-28 10:01:01.000*/,
+      dateAdd(
+          "millisecond",
+          60 * 1000 + 500,
+          Timestamp(1551348000, 500'999'999) /*2019-02-28 10:00:00.500*/));
+  EXPECT_EQ(
+      Timestamp(1551434400, 500'999'999) /*2019-03-01 10:00:00.500*/,
+      dateAdd(
+          "second",
+          60 * 60 * 24,
+          Timestamp(1551348000, 500'999'999) /*2019-02-28 10:00:00.500*/));
+  EXPECT_EQ(
+      Timestamp(1551434400, 500'999'999) /*2019-03-01 10:00:00.500*/,
+      dateAdd(
+          "minute",
+          60 * 24,
+          Timestamp(1551348000, 500'999'999) /*2019-02-28 10:00:00.500*/));
+  EXPECT_EQ(
+      Timestamp(1551434400, 500'999'999) /*2019-03-01 10:00:00.500*/,
+      dateAdd(
+          "hour",
+          24,
+          Timestamp(1551348000, 500'999'999) /*2019-02-28 10:00:00.500*/));
+  EXPECT_EQ(
+      Timestamp(1551434400, 500'999'999) /*2019-03-01 10:00:00.500*/,
+      dateAdd(
+          "day",
+          1,
+          Timestamp(1551348000, 500'999'999) /*2019-02-28 10:00:00.500*/));
+  EXPECT_EQ(
+      Timestamp(1585389600, 500'999'999) /*2020-03-28 10:00:00.500*/,
+      dateAdd(
+          "month",
+          12 + 1,
+          Timestamp(1551348000, 500'999'999) /*2019-02-28 10:00:00.500*/));
+  EXPECT_EQ(
+      Timestamp(1582884000, 500'999'999) /*2020-02-28 10:00:00.500*/,
+      dateAdd(
+          "quarter",
+          4,
+          Timestamp(1551348000, 500'999'999) /*2019-02-28 10:00:00.500*/));
+  EXPECT_EQ(
+      Timestamp(1582884000, 500'999'999) /*2020-02-28 10:00:00.500*/,
+      dateAdd(
+          "year",
+          1,
+          Timestamp(1551348000, 500'999'999) /*2019-02-28 10:00:00.500*/));
+
+  // Account for daylight saving. Daylight saving should not make any difference
+  // than normal dates. Daylight saving in US starts at 2021-03-14 02:00:00 PST
+  EXPECT_EQ(
+      Timestamp(1615716001, 999'999) /*2021-03-14 10:00:01.000*/,
+      dateAdd(
+          "millisecond",
+          60 * 60 * 1000 + 500,
+          Timestamp(1615712400, 500'999'999) /*2021-03-14 09:00:00.500*/));
+  EXPECT_EQ(
+      Timestamp(1615798800, 500'999'999) /*2021-03-15 09:00:00.500*/,
+      dateAdd(
+          "second",
+          60 * 60 * 24,
+          Timestamp(1615712400, 500'999'999) /*2021-03-14 09:00:00.500*/));
+  EXPECT_EQ(
+      Timestamp(1618390800, 500'999'999) /*2021-04-14 09:00:00.500*/,
+      dateAdd(
+          "minute",
+          60 * 24 * 31,
+          Timestamp(1615712400, 500'999'999) /*2021-03-14 09:00:00.500*/));
+  EXPECT_EQ(
+      Timestamp(1615798800, 500'999'999) /*2021-03-15 09:00:00.500*/,
+      dateAdd(
+          "hour",
+          24,
+          Timestamp(1615712400, 500'999'999) /*2021-03-14 09:00:00.500*/));
```

Comment:
Shouldn't the timestamp be before `02:00:00.000` in order to test impact due to daylight saving?
Probably add a test when daylight saving ends as well?

Reply:
Never mind. I see that the value is based on UTC.

Reply:
What about tests for TIMESTAMP_WITH_TIME_ZONE? Daylight saving should impact those?

Reply:
I'm adding support for the 3rd type TIMESTAMP_WITH_TIME_ZONE now as well as tests, and will patch this PR.